modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/compaction/Compactor.java

```diff
@@ -0,0 +1,372 @@
+        }
+    }
+
+    /**
+     * Waiting for delta files.
+     */
+    void waitDeltaFiles() {
+        try {
+            synchronized (mux) {
+                while (deltaFileCount.get() == 0 && !isCancelled()) {
+                    blockingSectionBegin();
+
+                    try {
+                        mux.wait();
+                    } finally {
+                        blockingSectionEnd();
+                    }
+                }
+            }
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+
+            isCancelled.set(true);
+        }
+    }
+
+    /**
+     * Adds the number of delta files to compact.
+     *
+     * @param count Number of delta files.
+     */
+    public void addDeltaFiles(int count) {
+        assert count >= 0;
+
+        if (count > 0) {
+            deltaFileCount.addAndGet(count);
+
+            synchronized (mux) {
+                mux.notifyAll();
+            }
+        }
+    }
+
+    /**
+     * Merges delta files with partition files.
+     */
+    void doCompaction() {
+        // Let's collect one delta file for each partition.
+        Queue<IgniteBiTuple<FilePageStore, DeltaFilePageStoreIo>> queue = filePageStoreManager.allPageStores().stream()
+                .flatMap(List::stream)
+                .map(filePageStore -> {
+                    DeltaFilePageStoreIo deltaFileToCompaction = filePageStore.getDeltaFileToCompaction();
+
+                    return deltaFileToCompaction == null ? null : new IgniteBiTuple<>(filePageStore, deltaFileToCompaction);
+                })
+                .filter(Objects::nonNull)
+                .collect(toCollection(ConcurrentLinkedQueue::new));
+
+        assert !queue.isEmpty();
+
+        updateHeartbeat();
+
+        int threads = threadPoolExecutor == null ? 1 : threadPoolExecutor.getMaximumPoolSize();
+
+        CompletableFuture<?>[] futures = new CompletableFuture[threads];
+
+        for (int i = 0; i < threads; i++) {
+            CompletableFuture<?> future = futures[i] = new CompletableFuture<>();
+
+            Runnable merger = () -> {
+                IgniteBiTuple<FilePageStore, DeltaFilePageStoreIo> toMerge;
+
+                try {
+                    while ((toMerge = queue.poll()) != null) {
+                        mergeDeltaFileToMainFile(toMerge.get1(), toMerge.get2());
+                    }
+                } catch (Throwable ex) {
+                    future.completeExceptionally(ex);
+                }
+
+                future.complete(null);
+            };
+
+            if (isCancelled()) {
+                return;
+            }
+
+            if (threadPoolExecutor == null) {
+                merger.run();
+            } else {
+                try {
+                    threadPoolExecutor.execute(merger);
+                } catch (RejectedExecutionException ignore) {
+                    // Run the task synchronously.
+                    merger.run();
```

Comment:
This is a copy of the code from 2.0, the rejection can be if it was not possible to add a worker due to a stop or in the process of shutting down. 

In theory, we won’t have this, because at the stop of the **Compactor** we say that we are canceling, we wait for all threads to complete and only then shutdown the thread pool. 

I will remove this code here and at the Checkpointer.